From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Francisco Boni <boboniboni@gmail.com>
Date: Tue, 14 Dec 2021 06:45:04 -0300
Subject: [PATCH] Add support for a stateless configuration file

Allow udisks to look at /usr/share/defaults/udisks2/udisks2.conf in
case the /etc/udisks/udisks2.conf file does not exist.

[insilications] Fixed location
[pmccarty] Rebased for udisks 2.8.4
[kprabhuv] Rebased for udisks 2.9.1
---
 src/udisksconfigmanager.c | 30 ++++++++++++++++++++++++------
 1 file changed, 24 insertions(+), 6 deletions(-)

diff --git a/src/udisksconfigmanager.c b/src/udisksconfigmanager.c
index ae612335..098fbaa0 100644
--- a/src/udisksconfigmanager.c
+++ b/src/udisksconfigmanager.c
@@ -145,7 +145,9 @@ parse_config_file (UDisksConfigManager         *manager,
                    GList                      **out_modules)
 {
   GKeyFile *config_file;
-  gchar *conf_filename;
+  gchar *conf_filename = NULL;
+  gchar *user_conf_filename;
+  gchar *system_conf_filename;
   gchar *load_preference;
   gchar *encryption;
   gchar *module_i;
@@ -154,17 +156,32 @@ parse_config_file (UDisksConfigManager         *manager,
   GError *l_error = NULL;
 
   /* Get modules and means of loading */
-  conf_filename = g_build_filename (G_DIR_SEPARATOR_S,
+  user_conf_filename = g_build_filename (G_DIR_SEPARATOR_S,
                                     manager->config_dir,
                                     PACKAGE_NAME_UDISKS2 ".conf",
                                     NULL);
 
-  udisks_debug ("Loading configuration file: %s", conf_filename);
+  system_conf_filename = g_build_filename (G_DIR_SEPARATOR_S,
+                                    "/usr/share/defaults/etc/udisks2",
+                                    PACKAGE_NAME_UDISKS2 ".conf",
+                                    NULL);
+
+  udisks_debug ("Loading configuration file: %s", user_conf_filename);
 
   /* Load config */
   config_file = g_key_file_new ();
   g_key_file_set_list_separator (config_file, ',');
-  if (g_key_file_load_from_file (config_file, conf_filename, G_KEY_FILE_NONE, &l_error))
+
+  if (g_key_file_load_from_file (config_file, user_conf_filename, G_KEY_FILE_NONE, NULL))
+  {
+        conf_filename = user_conf_filename;
+  }
+  else if (g_key_file_load_from_file (config_file, system_conf_filename, G_KEY_FILE_NONE, &l_error))
+  {
+        conf_filename = system_conf_filename;
+  }
+
+  if (conf_filename)
     {
       if (out_modules != NULL)
         {
@@ -228,7 +245,7 @@ parse_config_file (UDisksConfigManager         *manager,
     {
       if (l_error != NULL)
         {
-          udisks_warning ("Can't load configuration file %s: %s", conf_filename, l_error->message);
+          udisks_warning ("Can't load configuration file %s: %s", system_conf_filename, l_error->message);
           g_error_free (l_error);
         }
       else
@@ -239,7 +256,8 @@ parse_config_file (UDisksConfigManager         *manager,
     }
 
   g_key_file_free (config_file);
-  g_free (conf_filename);
+  g_free (user_conf_filename);
+  g_free (system_conf_filename);
 }
 
 static void
